File: src/web/types/system.rs
//! System monitor types for the htop-like panel.

use serde::Serialize;

/// Snapshot of system metrics, sent via SSE to the frontend.
#[derive(Serialize, Clone, Debug)]
pub struct SystemStats {
    /// Total physical memory in bytes.
    pub mem_total: u64,
    /// Used physical memory in bytes.
    pub mem_used: u64,
    /// Total swap in bytes.
    pub swap_total: u64,
    /// Used swap in bytes.
    pub swap_used: u64,
    /// Per-CPU usage percentages (0.0 – 100.0).
    pub cpu_usage: Vec<f32>,
    /// Overall CPU usage percentage (average across all cores).
    pub cpu_avg: f32,
    /// System uptime in seconds.
    pub uptime_secs: u64,
    /// System hostname.
    pub hostname: String,
    /// Load averages (1, 5, 15 min) — Linux/macOS only, empty on Windows.
    pub load_avg: [f64; 3],
    /// Top processes (limited to top 40).
    pub processes: Vec<ProcessInfo>,
    /// Total number of running processes.
    pub process_count: usize,
    /// Disk usage summaries.
    pub disks: Vec<DiskInfo>,
    /// Network interface stats.
    pub networks: Vec<NetworkInfo>,
}

/// Info about a single process.
#[derive(Serialize, Clone, Debug)]
pub struct ProcessInfo {
    pub pid: u32,
    pub name: String,
    /// CPU usage percentage for this process.
    pub cpu: f32,
    /// Memory usage in bytes.
    pub mem: u64,
    /// Process status string.
    pub status: String,
    /// Disk read bytes (total since process start).
    pub disk_read: u64,
    /// Disk write bytes (total since process start).
    pub disk_write: u64,
}

/// Info about a single disk/mount.
#[derive(Serialize, Clone, Debug)]
pub struct DiskInfo {
    pub name: String,
    pub mount: String,
    pub total: u64,
    pub used: u64,
    pub fs_type: String,
}

/// Info about a single network interface.
#[derive(Serialize, Clone, Debug)]
pub struct NetworkInfo {
    pub name: String,
    /// Total bytes received.
    pub rx_bytes: u64,
    /// Total bytes transmitted.
    pub tx_bytes: u64,
}

File: web-ui/src/api/system.ts
// ── Types ─────────────────────────────────────────────

export interface ProcessInfo {
  pid: number;
  name: string;
  cpu: number;
  mem: number;
  status: string;
  /** Disk read bytes (total since process start). */
  disk_read: number;
  /** Disk write bytes (total since process start). */
  disk_write: number;
}

export interface DiskInfo {
  name: string;
  mount: string;
  total: number;
  used: number;
  fs_type: string;
}

export interface NetworkInfo {
  name: string;
  rx_bytes: number;
  tx_bytes: number;
}

export interface SystemStats {
  mem_total: number;
  mem_used: number;
  swap_total: number;
  swap_used: number;
  cpu_usage: number[];
  cpu_avg: number;
  uptime_secs: number;
  hostname: string;
  load_avg: [number, number, number];
  processes: ProcessInfo[];
  process_count: number;
  disks: DiskInfo[];
  networks: NetworkInfo[];
}

// ── SSE stream ────────────────────────────────────────

/**
 * Connect to the system stats SSE stream.
 * Returns an EventSource that emits "system_stats" events with SystemStats JSON payloads.
 * Caller is responsible for closing the EventSource when done.
 */
export function connectSystemStatsStream(
  onStats: (stats: SystemStats) => void,
  onError?: (err: Event) => void,
): EventSource {
  const es = new EventSource("/api/system/stats/stream");

  es.addEventListener("system_stats", (e: MessageEvent) => {
    try {
      const stats: SystemStats = JSON.parse(e.data);
      onStats(stats);
    } catch {
      // ignore parse errors
    }
  });

  if (onError) {
    es.onerror = onError;
  }

  return es;
}
